lib/logstash/filters/elasticsearch/esql_executor.rb

```diff
@@ -0,0 +1,99 @@
+
+          @query = plugin.params["query"]
+          if @query.strip.start_with?("FROM") && !@query.match?(/\|\s*LIMIT/)
+            @logger.warn("ES|QL query doesn't contain LIMIT, adding `| LIMIT 1` to optimize the performance")
+            @query.concat(' | LIMIT 1')
+          end
+
+          query_params = plugin.params["query_params"] || {}
+          @named_params = query_params["named_params"] || []
+          @fields = plugin.params["fields"]
+          @tag_on_failure = plugin.params["tag_on_failure"]
+          @logger.debug("ES|QL query executor initialized with ", query: @query, named_params: @named_params)
+        end
+
+        def process(client, event)
+          resolved_params = @named_params&.any? ? resolve_parameters(event) : []
+          response = execute_query(client, resolved_params)
+          inform_warning(response)
+          process_response(event, response)
+          @plugin.decorate(event)
+        rescue => e
+          @logger.error("Failed to process ES|QL filter", exception: e)
+          @tag_on_failure.each { |tag| event.tag(tag) }
+        end
+
+        private
+
+        def resolve_parameters(event)
+          @named_params.map do |entry|
+            entry.each_with_object({}) do |(key, value), new_entry|
+              begin
+                resolved_value = event.get(value)
+                @logger.debug("Resolved value for #{key}: #{resolved_value}, its class: #{resolved_value.class}")
+                new_entry[key] = resolved_value
+              rescue => e
+                # catches invalid field reference
+                @logger.error("Failed to resolve parameter", key: key, value: value, error: e.message)
+                raise
+              end
+            end
+          end
+        end
+
+        def execute_query(client, params)
+          # debug logs  may help to check what query shape the plugin is sending to ES
+          @logger.debug("Executing ES|QL query", query: @query, params: params)
+          client.search({ body: { query: @query, params: params }, format: 'json', drop_null_columns: true }, 'esql')
+        end
+
+        def process_response(event, response)
+          columns = response['columns'].freeze
+          values = response['values'].freeze
+          if values.nil? || values.size == 0
+            @logger.debug("Empty ES|QL query result", columns: columns, values: values)
+            return
+          end
+
+          # this shouldn't never happen but just in case not crash the plugin
+          if columns.nil? || columns.size == 0
+            @logger.error("No columns exist but received values", columns: columns, values: values)
+            return
+          end
+
+          # TODO: do we need to set `total_hits` to target?
+          #   if not, how do we resolve conflict with existing es-input total_hits field?
+          #   FYI: with DSL it stores in `[@metadata][total_hits]`
+          event.set("[@metadata][total_hits]", values.size)
```

Comment:
Questions: I have made same with DSL but I would support:
- renaming this to `total_values`
- renaming seems breaking change (as my understanding we are changing interface with ES not event shape), how do we resolve conflict with es-input's `[@metadata][total_hits]`? - maybe set to target if defined?

Reply:
@yaauie WDYT if we set to `[@metadata][total_values]` under the `target`?

Reply:
> Users have never been able to make ES|QL queries before, ...

Ah sorry, I meant renaming `total_hits` (still I don't like using hits 👎 ) of the DSL executor. But I think here it is clear that we introduce `[@metadata][total_values]`.

> I don't think we should conflate the metadata with our newly-introduced target. ..

Yup, thank you. Since we don't have `metadata_target`, this was my question to confirm.